service/s3/s3crypto/crypto_mode.go

```diff
@@ -0,0 +1,138 @@
+package s3crypto
+
+import (
+	"encoding/base64"
+	"io"
+	"strconv"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+)
+
+// CryptoMode is an abstraction layer that deals with encryption of
+// the contents and which key provider to use.
+type CryptoMode interface {
+	EncryptContents(io.Reader) (io.Reader, error)
+	DecryptContents(KeyProvider, io.ReadCloser) (io.ReadCloser, error)
+	GetKeyProvider() KeyProvider
+	CipherDataMetadata
+}
+
+// CipherDataMetadata  is used for when populating the envelope details upon
+// encryption.
+type CipherDataMetadata interface {
+	GetCipherName() string
+	GetTagLen() string
+}
+
+// DecryptMode is meant to used only in reading objects from s3
+type DecryptMode interface {
+	DecryptContents(KeyProvider, io.ReadCloser) (io.ReadCloser, error)
+	GetKeyProvider() KeyProvider
+}
+
+func modeForEnvelope(env *Envelope, cfg Config) (DecryptMode, error) {
+	kp, err := keyProviderForEnvelope(env, cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	err = DecodeMeta(env, kp)
+	if err != nil {
+		return nil, err
+	}
+
+	kp.SetKey([]byte(env.CipherKey))
+	kp.SetIV([]byte(env.IV))
+
+	cek, err := cekForEnvelope(env, kp)
```

Comment:
If `cekForEnvelope` is updated to just take the algorithm and not the full envelope won't need to pass in env.